Import in Signup: Add preview step to import-onboarding flow

andfinally opened the pull request:
#### Changes proposed in this Pull Request

* This PR adds a "preview" step to the import signup flow. It's intended to set expectations about what the user can expect at the end of the import process. We plan to show it after someone has entered the URL of a site they want to import. We should have detected the site type by that stage.
* Based on the type of site detected, we show a basic list of features we'll import, and a sentence about what we can't import.

![image](https://user-images.githubusercontent.com/1647564/62698632-7f836180-b9d5-11e9-9246-27b223f144ac.png)

* At this stage, we'll show the icon for the service (for example GoDaddy) with the site title and URL instead of the site's favicon. This is because not all favicons are big enough to show here. We'll need to deal with that in a separate PR.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply the PR and go to http://calypso.localhost:3000/start/import-onboarding/, or use `calypso.live`.
* Click the `Already have a website?` link on the `site-type` step to go to the `from-url` step.
* Enter the URL of a GoDaddy or Wix site to import from, for example https://omowale.org or https://www.totalsportswomensrace.com.
* You should see the preview. Clicking the "Continue" button should take you to the `domains` step and ultimately to the GoDaddy or Wix importer, where the site preview should load. (We'll look at skipping that preview in another PR.)
* Check this new step only appears in the `import-onboarding` flow and not in the existing import flow from `/start/import`.

* Start again at `/start/import-onboarding/` and select one of the site types, like "Blog".
* You should to the next step in the main onboarding flow, `/start/site-topic-with-preview`.


Fixes https://github.com/Automattic/samus-private/issues/198

gwwar commented:
Is there more context for this change?

gwwar commented on apps/full-site-editing/full-site-editing-plugin/full-site-editing/class-full-site-editing.php:
```diff
@@ -191,12 +191,12 @@ public function register_blocks() {
-					],
-				],
+					),
+				),
```
Why all the janitorial changes here?

gwwar replied:
There doesn't appear to be any other changes to the Full Site Editing plugin? Do you mind dropping the related commits?

vindl replied:
+1 For dropping these changes, we've been using array shorthand syntax throughout the FSE plugin on purpose. For context, our minimum plugin requirements are PHP 5.6+.

andfinally replied:
Thanks @gwwar and @vindl, that change should not have been included in this draft – it's now removed.

a8ci18n commented on client/signup/steps/import-preview/index.jsx:
```diff
@@ -0,0 +1,228 @@
+				supported: [
+					translate( 'Posts' ),
+					translate( 'Pages' ),
+					translate( 'Images and other media' ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Finding Free Images and other Media' )` *ES Score: 6*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

mattsherman commented on client/my-sites/importer/logos/godaddy-gocentral.jsx:
```diff
@@ -5,11 +5,14 @@
+} ) {
 	return (
 		<img
 			// eslint-disable-next-line wpcalypso/jsx-classname-namespace
```
Do we still need the `eslint-disable-next-line` here?

andfinally replied:
The linter is happy with it removed, so I've taken it out.

mattsherman commented on client/my-sites/importer/logos/medium.jsx:
```diff
@@ -5,10 +5,10 @@
-export default () => (
+export default ( { className = 'medium-logo social-logo importer__service-icon' } ) => (
 	<svg
 		/*eslint-disable wpcalypso/jsx-classname-namespace */
```
Do we still need these `eslint-disable`s? If so, should replace with `eslint-disable-next-line` to tidy things up a bit.

andfinally replied:
The linter complains when I try to remove this one, so I've changed it to a single line comment.

mattsherman commented on client/my-sites/importer/importer-logo.jsx:
```diff
@@ -19,34 +19,37 @@ import SocialLogo from 'components/social-logo';
+const ImporterLogo = ( { icon, className = 'importer__service-icon' } ) => {
+	/* eslint-disable wpcalypso/jsx-classname-namespace */
+
+	if ( 'blogger' === icon ) {
```
Why this special handling for `blogger`?

jonathansadowski replied:
Instead of `icon={ importSiteEngine }`, we can get this mapping (that the icon for `blogger` is `blogger-alt`) from `client/lib/importer/importer-config`. It doesn't seem right to map engine to the actual icon configured for the engine within this component.

andfinally replied:
Meeps, this was because we have two icons for Blogger, and use the version called `blogger-alt`. Thanks Jon, now changed.

mattsherman commented on client/my-sites/importer/importer-logo.jsx:
```diff
@@ -19,34 +19,37 @@ import SocialLogo from 'components/social-logo';
 
 	if ( 'wix' === icon ) {
-		return <WixLogo />;
+		return <WixLogo className={ className } />;
```
Why this change? (and similar ones for other services)

andfinally replied:
I was hoping to keep the markup tidy by giving these components classes that were consistent with their parents in the import preview. I also thought it would help avoid having to override existing CSS. But the change made the components less simple, and it seems the linter has no problem with me using the existing classnames, so I've removed these params, thanks!

mattsherman commented on client/my-sites/importer/importer-logo.jsx:
```diff
@@ -19,34 +19,37 @@ import SocialLogo from 'components/social-logo';
 
 	if ( 'godaddy-gocentral' === icon ) {
-		return <GoDaddyGoCentralLogo size={ 48 } />;
+		return <GoDaddyGoCentralLogo size={ 48 } className={ className } />;
```
Why is GoDaddy the only one that has a size specified?

andfinally replied:
I'm not sure why a size param was implemented in this particular icon but not in the others – I've left it in and given it a default rather than lose the extra flexibility it brings.

mattsherman commented on client/my-sites/importer/logos/godaddy-gocentral.jsx:
```diff
@@ -5,11 +5,14 @@
 		<img
 			// eslint-disable-next-line wpcalypso/jsx-classname-namespace
-			className="godaddy-gocentral social-logo importer__service-icon"
+			className={ className }
```
Why this change?

andfinally replied:
Please see response in `importer-logo.jsx`.

mattsherman commented:
@andfinally In addition to addressing the comments, please make sure to rebase this against the latest in master to make sure we pick up the latest changes made in #35356.

I'll give this another spin once it is ready!

andfinally commented:
I'm not sure why we were skipping the domains step, but we were already doing this in the `import` and `import-onboarding` flows – until recently, when https://github.com/Automattic/wp-calypso/pull/35175 was merged, and now we're getting the step again.

Thanks for spotting the issue with the logos on the importer page! With the latest treatment of the icon CSS classes this is now fixed.

I've rebased the branch.

mattsherman commented:
Thanks for making the updates @andfinally. Once #35409 is merged, please rebase against master so that we can see how the flow looks with that in place. I'll give this a final review/test then!

andfinally commented:
Thanks @mattsherman, the branch is now rebased.

jonathansadowski commented on client/signup/steps/import-preview/style.scss:
```diff
@@ -0,0 +1,69 @@
+.import-preview__card {
+	max-width: 360px;
+	padding: rem( 16px );
```
I believe most of the steps have a `border-radius` of `3px`. Can we add that here for consistency?

andfinally replied:
Thanks Jon, now added.

jonathansadowski commented on client/signup/steps/import-preview/index.jsx:
```diff
@@ -0,0 +1,208 @@
+		if ( isEmpty( importSiteUrl ) ) {
+			// The user has an import file and told us the source
+			const engine = importSiteEngine === 'wordpress' ? 'WordPress' : startCase( importSiteEngine );
+			cardTitle = 'Importing from ' + engine;
```
Can you wrap this in a translate? And can we use the config to get the engine title instead of calculating it from engine?

andfinally replied:
Thanks, that's better!

andfinally commented on client/signup/steps/import-preview/index.jsx:
```diff
@@ -0,0 +1,209 @@
+			const importerData = importerConfig()[ importSiteEngine ];
+			cardTitle = translate( 'Importing from ' ) + importerData.title;
+		} else {
+			trimmedSiteUrl = importSiteUrl.replace( /^https?:\/\//i, '' ).replace( /\/$/, '' );
```
Too tired to make a combined regex pattern work right now, so I've done a double `replace`.

jonathansadowski replied:
If you want to switch it, this should work, but I think it's also fine as two replaces:
```suggestion
			trimmedSiteUrl = importSiteUrl.replace( /^https?:\/\/(.*?)\/$/, '$1' );
```

andfinally replied:
I found I had to make the outer subpatterns optional, but now it works, thanks Jon.

jonathansadowski commented on client/signup/steps/import-preview/index.jsx:
```diff
@@ -0,0 +1,209 @@
+		if ( isEmpty( importSiteUrl ) ) {
+			// The user has an import file and told us the source
+			const importerData = importerConfig()[ importSiteEngine ];
+			cardTitle = translate( 'Importing from ' ) + importerData.title;
```
How about something like this?

```suggestion
			cardTitle = translate( 'Importing from %(importerName)s', {
				args: { importerName: importerData.title },
			} );
```

This makes the string more translatable in languages where the translation would need to put the engine name first.

andfinally replied:
Ahh much better, thanks!

p-jackson commented on client/signup/steps/import-preview/index.jsx:
```diff
@@ -0,0 +1,209 @@
+	render() {
+		const { flowName, positionInFlow, stepName, translate } = this.props;
+
+		const headerText = translate( "Here's what we'll import to your new" + ' site.' );
```
Maybe tidy up this string concat? My guess is it was just prettier going haywire.

andfinally replied:
Thanks Philip, I think it was - now tidied.

andfinally commented:
Thanks for having a look @p-jackson! 😄 The behaviour on refresh is intentional – some of the flows, including this one, have the `disallowResume` parameter set, which returns you to the first step on refresh (the site type page in your case, since you're probably logged in). I'm not sure why we're doing this on some flows, but I suspect it's to ensure they hang on to all the data they need.

Your point about the flow from the main options on the site type list has saved us from introducing a regression, thank you! 🙏 I've added a condition in the `site-type` component that switches the flow back to onboarding when you select one of the main site types. It does feel a bit brittle checking for a particular flow in this condition, but I'm not sure there's a better way at this stage.
